[WIP] Start of work to redo the threading and forking split

jimi-c opened the pull request:
This refactors the way we handle process management by introducing the ProcessModel* classes. Right now, these will remain hard-coded as there's no real plans to add others in the future beyond threading/forking.

As a WIP, this PR enables threading by default, however when merged forking will be the default as it is today. Threading will have to be enabled via a config change (environment variable or ansible.cfg).

jimi-c committed:
Start of work to redo the threading and forking split

ci_complete
ci_coverage

jimi-c commented:
!jimi_skip

sivel commented on lib/ansible/executor/process/threading.py:
```diff
@@ -0,0 +1,170 @@
+    def put_result(self, data):
+        try:
+            self._final_q_lock.acquire()
+            self._final_q.appendleft(data)
```
I don't think we want to `appendleft` here.  Otherwise that would make this queue act like LIFO since we use `popleft` in `get_results`

sivel commented on lib/ansible/executor/process/threading.py:
```diff
@@ -0,0 +1,170 @@
+            self._job_queue_lock.release()
+        return data
+
+    def put_task(self, data):
```
Just as a note here, we call this `put_task` but getting the data from the queue we call it `get_job` which seems inconsistently named.

sivel replied:
Also, instead of using `appendleft` here and `pop` above, we could `append`, and `popleft`, it doesn't really matter.  Although I think we should order all of our queues so they basically operate the same.

sivel commented on lib/ansible/config/base.yml:
```diff
@@ -571,6 +571,16 @@ DEFAULT_FORKS:
   type: integer
+DEFAULT_PROCESS_MODEL:
+  name: Sets the Ansible process model
+  default: "threading"
```
I know the default was likely set to `threading` to get tests to run, but we need to remember to flip this.

sivel commented on lib/ansible/executor/process/forking.py:
```diff
@@ -37,17 +39,103 @@
+        self._cur_worker = 0
+        self._workers = [None for i in range(num)]
+
+    def put_task(self, data):
```
Can we explode the args to `put_task`?  I'm not a huge fan of passing it as one arg, tho then explode it again.  individual args seems a better idea.

sivel commented on lib/ansible/executor/process/forking.py:
```diff
@@ -37,17 +39,103 @@
+        # create the worker results queue, which is where all workers
+        # send their results. The results thread reads from this in the
+        # background and copies the results to the final_q deque object
+        self._worker_results_q = multiprocessing.Queue()
```
Need to make sure we add the `try/except` back around this to provide a friendlier error when `/dev/shm` is not available.  It was in the removed code.

sivel commented:
A general comment, `Ctrl-c` doesn't seem to work with threading.  I cannot abort the playbook run.

jimi-c committed:
Fix ctrl+c handling for the new forking/threading models

sivel commented on lib/ansible/executor/process/threading.py:
```diff
@@ -0,0 +1,171 @@
+        for i in range(num):
+            w_thread = threading.Thread(target=run_worker, args=(self, self._shared_loader_obj))
+            w_thread.start()
+            w_lock = threading.Lock()
```
This lock is not used anywhere, and we removed a similar lock from forking, so we need to remove it here too.

dw commented:
Is there any chance of the plugins/loader.py patches landing for 2.8? They are fairly low impact, and useful elsewhere regardless of the remaining refactor

Also I'm worried https://github.com/ansible/ansible/pull/45572 may need reverted with respect to this PR, it shares the Templar across threads:

> It does not seem possible for concurrent calls into one (non-copied) VariableManager instance, and so it need not be reentrant. If that became a requirement, serializing its or Templar's entry points would be fine, as it's so CPU-heavy other threads will only fight with it for the GIL anyway.

jimi-c committed:
Cleaning up problems with threading after rebasing

jimi-c committed:
Make the process model a proper plugin

* Also reverts the default model back to forking

jimi-c committed:
Expand args passed to put_job for process models

jimi-c committed:
Wrap use of multiprocess.Queue in try/except

This was in the original TQM use to protect against the possibility
that /dev/shm was not available.

jimi-c committed:
Remove unused lock from threading worker setup

sivel commented on lib/ansible/plugins/process/threading.py:
```diff
@@ -0,0 +1,177 @@
+        self._cur_worker = 0
+        self._workers = []
+        for i in range(num):
+            w_thread = threading.Thread(target=run_worker, args=(self, self._shared_loader_obj))
```
Should we strive to make `threading` and `forking` work more similarly?  Here we create a pool of "workers" to handle the work. In `forking` we don't re-allocate a pool of workers, we spawn and reap those workers repeatedly over the lifetime of the execution.

sivel replied:
Sounds like the answer is "No".

jimi-c replied:
As discussed, you can't keep a pool of forked workers because you have to serialize things sent to them, which was a HUGE slowdown in the original 2.0 beta.

jimi-c committed:
Instantiate the process model class in run() for TQM, not init

sivel commented on lib/ansible/plugins/process/threading.py:
```diff
@@ -0,0 +1,177 @@
+        data = None
+        try:
+            data = self._job_queue.popleft()
+        except:
```
Insert `self._tqm.send_callback('v2_runner_on_start', data[0], data[1])` here

sivel commented on lib/ansible/plugins/process/threading.py:
```diff
@@ -0,0 +1,177 @@
+    def put_job(self, host, task, play_context, task_vars):
+        try:
+            self._job_queue.append((host, task, play_context, task_vars))
+            self._tqm.send_callback('v2_runner_on_start', host, task)
```
Nuke this line

jimi-c committed:
Adding documentation and fixing up plugin-related areas

* Also fixes a bug in which the callback for tasks being started in
  threads was when the task data was put on the queue, not when it was
  popped by a worker and actually started.

sivel commented on lib/ansible/plugins/loader.py:
```diff
@@ -831,3 +831,11 @@ def _load_plugin_filter():
+process_loader = PluginLoader(
+    'ProcessModel',
+    'ansible.plugins.process',
+    C.DEFAULT_PROCESS_MODEL,
```
You need a new config for `DEFAULT_PROCESS_PLUGIN_PATH`, and reference that here.  Instead of just inserting the plugin name here.

sivel commented on lib/ansible/plugins/process/__init__.py:
```diff
@@ -0,0 +1,98 @@
+__all__ = ["ProcessModelBase", "keyboard_interrupt_event", "ResultsSentinel"]
+
+
+class ResultsSentinel:
```
FWIW, we have a sentinel available in `ansible.utils.sentinel` now. Not sure if it's worth using that instead.

sivel commented on lib/ansible/plugins/process/forking.py:
```diff
@@ -40,25 +53,117 @@
+__all__ = ['ProcessModel',]
+
+
+class ProcessModel(ProcessModelBase):
```
Not sure if it would be useful for something later, but should we have a class attr with the name of the plugin in it?  Some plugin types do this others don't.

bcoca replied:
_load_name gets appended by loader itself

sivel commented on lib/ansible/plugins/process/threading.py:
```diff
@@ -0,0 +1,189 @@
+        data = None
+        try:
+            data = self._job_queue.popleft()
+            self._tqm.send_callback('v2_runner_on_start', host, task)
```
`host, task` should instead be `data[0], data[1]`

jimi-c committed:
Create a plugin path config value for process plugins

jimi-c committed:
Fix threading callback use of variables

bcoca commented on lib/ansible/plugins/process/__init__.py:
```diff
@@ -0,0 +1,98 @@
+    keyboard_interrupt_event.set()
+
+
+class ProcessModelBase(with_metaclass(ABCMeta, object)):
```
you might want to inherit from AnsiblePlugin, among other things, this will give you 'option processing' for free, not needed in current plugins but might want for future

bcoca commented on lib/ansible/plugins/process/forking.py:
```diff
@@ -19,9 +19,22 @@
+        - This process model creates workers using the Python multiprocessing library. Workers are created
+          on-demand as jobs are queued up, up to the maximum number of forks allowed.
+    author: ansible (@core)
+    version_added: 2.8
```
version added requires quotes, otherwise it will be processed as number

bcoca commented on lib/ansible/executor/shared_plugin_loader.py:
```diff
@@ -18,3 +18,22 @@
+from ansible.plugins.loader import action_loader, connection_loader, filter_loader, lookup_loader, module_loader, test_loader
+
+
+class SharedPluginLoaderObj(with_metaclass(ABCMeta, object)):
```
still not sure why we need this, just `from ansible.plugins import loader as shared_loader` seems to do same thing?

bcoca commented on lib/ansible/config/base.yml:
```diff
@@ -613,6 +613,24 @@ DEFAULT_FORKS:
   ini:
   - {key: forks, section: defaults}
   type: integer
+DEFAULT_PROCESS_MODEL:
```
DEFAULT_ is misnomer (originally true but not so since 0.2) I stopped using it a while back, we should really drop it from any new settings

jimi-c committed:
Adding testing for threading

acozine commented on lib/ansible/constants.py:
```diff
@@ -104,7 +104,7 @@ def __getitem__(self, y):
 DEFAULT_SU_PASS = None
 # FIXME: expand to other plugins, but never doc fragments
-CONFIGURABLE_PLUGINS = ('become', 'cache', 'callback', 'cliconf', 'connection', 'httpapi', 'inventory', 'lookup', 'shell')
+CONFIGURABLE_PLUGINS = ('become', 'cache', 'callback', 'cliconf', 'connection', 'httpapi', 'inventory', 'lookup', 'shell', 'process')
```
```suggestion
CONFIGURABLE_PLUGINS = ('become', 'cache', 'callback', 'cliconf', 'connection', 'httpapi', 'inventory', 'lookup', 'process', 'shell' )
```

acozine commented on docs/docsite/Makefile:
```diff
@@ -26,7 +26,7 @@ ifdef PLUGINS
 endif
 
-DOC_PLUGINS ?= become cache callback cliconf connection httpapi inventory lookup shell strategy vars
+DOC_PLUGINS ?= become cache callback cliconf connection httpapi inventory lookup shell strategy vars process
```
```suggestion
DOC_PLUGINS ?= become cache callback cliconf connection httpapi inventory lookup process shell strategy vars
```

acozine commented on docs/docsite/rst/plugins/plugins.rst:
```diff
@@ -25,6 +25,7 @@ This section covers the various types of plugins that are included with Ansible:
    shell
    strategy
    vars
+   process
```
To preserve the alphabetization, `process` should slot in under `lookup` and above `shell`.

acozine commented on docs/docsite/rst/plugins/process.rst:
```diff
@@ -0,0 +1,50 @@
+
+Process plugins control the way Ansible creates workers to run tasks on hosts. Currently, the specified process model is global and cannot be changed once Ansible starts.
+
+By default, Ansible ships with two process models: :ref:`forking<forking_process>` and :ref:`threading<threading_process>`.
```
Some information here about why someone might want to change their process model - not necessarily pointers on which one to choose, but more general considerations. For example (and I'm making this up, but it's a semi-educated guess): "Selecting the right process plugin can improve Ansible's performance in your environment. You should match the process model to your available resources." Or maybe: "Process models have security implications. You should select a process plugin with care to ensure secure operations." Whatever the specifics, an indication of why the user might care will improve this page.

mattclay committed:
Short-term solution for testing threading in CI.

A longer term solution should move the decision to run with
threading or forking into the tests themselves, instead of being
dependent on the configuration of the test environment.

ci_complete

acozine reviewed (commented):
I added a couple of suggestions for improving the docs portion of this PR.

jimi-c committed:
Re-add locks to plugin loader for thread safety

jimi-c committed:
Revert "Adding testing for threading"

This reverts commit 4c66fa35cfa562480dbc223e0fa22be9e01f043d.

jimi-c committed:
Fix typo in indentation

ci_complete

jimi-c committed:
Fixing things up for tests

!ci_complete

jimi-c committed:
Improving ctrl+c handling for process models

!ci_complete

emousse commented:
Hi! 
Can you please tell me if you are still working on this branch ?

jimi-c commented:
@emousse not really. A few things popped up regarding non-thread-safe methods we use in python, and I cannot confidently say those will ever be addressed. As such I've turned my attention to alternative ways of improving the performance of the core engine. I'm going to go ahead and close this WIP now as it's quite out of date.
